include/rmm/mr/device/arena_memory_resource.hpp

```diff
@@ -0,0 +1,287 @@
+  arena_memory_resource& operator=(arena_memory_resource const&) = delete;
+
+  /**
+   * @brief Queries whether the resource supports use of non-null CUDA streams for
+   * allocation/deallocation.
+   *
+   * @returns bool true.
+   */
+  bool supports_streams() const noexcept override { return true; }
+
+  /**
+   * @brief Query whether the resource supports the get_mem_info API.
+   *
+   * @return bool false
+   */
+  bool supports_get_mem_info() const noexcept override { return false; }
+
+ private:
+  using arena      = detail::arena::arena<Upstream>;
+  using read_lock  = std::shared_lock<std::shared_timed_mutex>;
+  using write_lock = std::lock_guard<std::shared_timed_mutex>;
+
+  /**
+   * @brief Allocates memory of size at least `bytes`.
+   *
+   * The returned pointer has at least 256B alignment.
+   *
+   * @throws `std::bad_alloc` if the requested allocation could not be fulfilled
+   *
+   * @param bytes The size in bytes of the allocation
+   * @param stream The stream to associate this allocation with
+   * @return void* Pointer to the newly allocated memory
+   */
+  void* do_allocate(std::size_t bytes, cudaStream_t stream) override
+  {
+    if (bytes <= 0) return nullptr;
+
+    bytes = rmm::detail::align_up(bytes, allocation_alignment);
+    RMM_EXPECTS(
+      bytes <= get_maximum_allocation_size(), rmm::bad_alloc, "Maximum allocation size exceeded");
+
+    if (arena::handles_size(bytes)) {
+      return get_arena(stream).allocate(bytes);
+    } else {
+      RMM_ASSERT_CUDA_SUCCESS(cudaStreamSynchronize(stream));
+      return upstream_mr_->allocate(bytes, cudaStreamLegacy);
+    }
+  }
+
+  /**
+   * @brief Deallocate memory pointed to by `p`.
+   *
+   * @param p Pointer to be deallocated
+   * @param bytes The size in bytes of the allocation. This must be equal to the
+   * value of `bytes` that was passed to the `allocate` call that returned `p`.
+   * @param stream Stream on which to perform deallocation
+   */
+  void do_deallocate(void* p, std::size_t bytes, cudaStream_t stream) override
+  {
+    if (p == nullptr || bytes <= 0) return;
+
+    bytes = rmm::detail::align_up(bytes, allocation_alignment);
+    if (arena::handles_size(bytes)) {
+      if (!get_arena(stream).deallocate(p, bytes, stream)) {
+        deallocate_across_arenas(p, bytes, stream);
+      }
+    } else {
+      RMM_ASSERT_CUDA_SUCCESS(cudaStreamSynchronize(stream));
+      upstream_mr_->deallocate(p, bytes, cudaStreamLegacy);
+    }
+  }
+
+  /**
+   * @brief Deallocate memory pointed to by `p` that was allocated in a different arena.
+   *
+   * @param p Pointer to be deallocated
+   * @param bytes The size in bytes of the allocation. This must be equal to the
+   * value of `bytes` that was passed to the `allocate` call that returned `p`.
+   * @param stream Stream on which to perform deallocation
+   */
+  void deallocate_across_arenas(void* p, std::size_t bytes, cudaStream_t stream)
```

Comment:
I find the term `across` ambiguous in this usage. Initially I thought it meant "allocate a block that spans more than one arena", but it actually means "deallocate a block that was allocated *by* a different arena".

I think a different term should be used. e.g. `deallocate_from_other_arena`, or possibly `deallocate_unowned`.

Reply:
Done.